doc: add enzyme integration demo + blog post (#542)
pasteurlabs/tesseract-core: docs/blog/2026-06-29-enzyme-lfortran-autodiff.md

What if you could backpropagate through existing Fortran, C, or C++ simulation code, embed it into JAX and torch, and use it as a high-performance differentiable physics engine? Turns out, you can — if you're brave enough...

Decades of validated physics code in CFD, climate, aerospace, and nuclear sit behind a wall that modern ML pipelines can't cross, because they don't expose gradients. The usual answer is to rewrite it all in JAX or PyTorch. The alternative we explore here is to leave the code where it is and get exact gradients out anyway, thanks to some LLVM-level magic. [Enzyme](https://enzyme.mit.edu/) applies autodiff at the LLVM IR level, so we can differentiate any code that compiles to LLVM.

All we need is to duct-tape [LFortran](https://lfortran.org/), LLVM, and Enzyme together, point the result at a Fortran thermal solver, and get exact gradients out the other end. From there, [Tesseract](https://github.com/pasteurlabs/tesseract-core) (whose blog you're reading) wraps the result as a custom [JAX](https://jax.readthedocs.io/en/latest/) primitive, so a Fortran solver becomes a differentiable layer in arbitrary JAX code.

> This is all pretty experimental, so you may have to spend some time chasing a gradient that returned NaN and manually compare LLVM IR diffs to make it work. But the gradients come out the other end of the entire multi-step time loop matching the analytic answer. And it's amazing to see that a stack combining some of the oldest and newest technologies can work together at all.

Reviewer: ```suggestion
This is all pretty experimental, so you may have to spend some time chasing a gradient that returned NaN and manually compare LLVM IR diffs to make it work. But in this case the gradients come out the other end of the entire multi-step time loop matching the analytic answer. And it's amazing to see that a stack combining some of the oldest and newest technologies can work together at all.
```

Author: ```suggestion
This is all pretty experimental, so you may have to spend some time chasing a gradient that returned NaN and manually compare LLVM IR diffs to make it work. But after some work, the gradients that come out the other end of the entire multi-step time loop match an analytic answer. And it's amazing to see that a stack combining some of the oldest and newest technologies can work together at all.
```

> Let's start at the beginning. What follows is the full walkthrough, including the compilation pipeline, the sharp edges we hit, and an inverse problem that made the whole effort worth it (and that wouldn't be possible without AD).

Reviewer: reducing AI smell
```suggestion
Let's start at the beginning. What follows is the full walkthrough, including the compilation pipeline, the sharp edges we hit, and application on a challenging inverse problem (that wouldn't be possible without AD).
```

        ! Harmonic-mean conductivity at cell faces
        kx_east = 2.0d0 * (k0 + k1*T_c) * (k0 + k1*T_e) &
                  / ((k0 + k1*T_c) + (k0 + k1*T_e))
        kx_west = 2.0d0 * (k0 + k1*T_c) * (k0 + k1*T_w) &
                  / ((k0 + k1*T_c) + (k0 + k1*T_w))
        ! ... (ky_north, ky_south similarly)

        flux_x = (kx_east*(T_e - T_c) - kx_west*(T_c - T_w)) / (dx*dx)
        flux_y = (ky_north*(T_nn - T_c) - ky_south*(T_c - T_s)) / (dy*dy)

        T_new(idx) = T_c + dt/(rho*cp) * (flux_x + flux_y + Q(idx))
      end do
    end do
    ! ... (boundary conditions, swap T_cur <- T_new)
  end do
```

Boundary conditions are mixed: Dirichlet (hot wall at the bottom), convection/Robin (top), and insulated/Neumann (sides). The faces use harmonic-mean conductivity, the standard trick for flux continuity across cells with different $k$.

For this setup, there are no easy shortcuts to get derivatives: because $k(T)$ is nonlinear, the stencil coefficients depend on the current temperature field, so the Jacobian changes at every time step. Hand-coding the adjoint through that, and re-deriving it every time the forward code changes, is the kind of work AD exists to delete.

> This is vanilla Fortran with no annotations or AD-aware constructs, but even getting to "vanilla" took some massaging. Allocatables, array intrinsics, and bounds checking all compile down to runtime calls (`_lfortran_malloc` and friends) that Enzyme can't see through, so they had to go. Work arrays get passed in pre-allocated from C, and we compile with `--no-array-bounds-checking`. For a 220-line solver that's an hour of work; for a large legacy codebase, it's an open question.

Reviewer: Strong AI smell here.

```suggestion
Note that Enzyme can't handle allocatables, array intrinsics, and bounds checking, As such, work arrays get passed in pre-allocated from C, and we compile with `--no-array-bounds-checking`. For a 220-line solver that's an hour of work; for a large legacy codebase, it's an open question.
```

Author: Enzyme doesn't know about those Fortran constructs, this is all LFortran's regime.

```suggestion
This is vanilla Fortran with no annotations or AD-aware constructs, but even getting to "vanilla" took some massaging. In LFortran, allocatables, array intrinsics and bounds checking all compile down to runtime calls (`_lfortran_malloc` and friends) that Enzyme can't see through, so we have to avoid them here. Instead, work arrays get passed in pre-allocated from C, and we compile with `--no-array-bounds-checking`. For a 220-line solver that's an hour of work; for a large legacy codebase, it's an open question.
```

## The Enzyme compilation pipeline

This is where it gets technical, so if you're not into the low-level details, feel free to [skip to the next section](#does-it-actually-work) where we benchmark the gradients.

> Enzyme works at the LLVM IR level, not the source level, so anything that compiles to LLVM IR (C, C++, Rust, Fortran) can be differentiated. The chain is six `opt`/`clang` invocations, surprisingly straightforward as long as you're comfortable inspecting IR when things go wrong.

Reviewer: ```suggestion
Enzyme works at the LLVM IR (intermediate representation) level, not the source level, so anything that compiles to LLVM IR (C, C++, Rust, Fortran) can be differentiated. The chain is six `opt`/`clang` invocations, surprisingly straightforward as long as you're comfortable inspecting IR when things go wrong.
```

Reviewer: Also not the "as long as you're comfortable inspecting IR when things go wrong." caveat seems overly confident, if you disagree qualify this by a statement such as "no more complicated than a jaxpr" or "trivial with the help of an LLM" or "see below" with a. link.

Author: Yea I like see below.

Reviewer: I don't think this change was made.

The six-step compilation pipeline, from Fortran source to a differentiable shared library. LFortran lowers the solver to LLVM IR, a thin C wrapper is linked in, Enzyme synthesizes the derivative code, and a final optimization pass emits a single `.so` exposing forward, JVP, and VJP entry points. The whole chain runs in about 30 seconds during `tesseract build`.
```

Most of those steps are plumbing: lower the Fortran to IR, compile a thin C wrapper, `llvm-link` the two, run Enzyme, optimize, emit a `.so`. The three that actually matter are the LFortran IR dump, the pre-Enzyme `-O1` cleanup, and the Enzyme pass itself.

```bash
# The compilation pipeline: Fortran source → differentiated shared library

# Fortran → LLVM IR (clean enough for Enzyme to trace)
$ lfortran --show-llvm --no-array-bounds-checking thermal_2d.f90 > thermal_2d.ll

# Mild cleanup BEFORE the AD pass — -O1, deliberately not -O3 (see below)
$ opt -O1 -S thermal_2d.ll -o thermal_2d_opt.ll

# ... link in the C wrapper etc ...

# Synthesize the derivative
$ opt --load-pass-plugin=LLVMEnzyme-19.so -passes=enzyme combined.ll -o ad.ll

# ... final optimization and emit a shared library ...
```

> A couple of decisions along the way shaped how the whole thing holds together, though, and they're worth dwelling on.

Reviewer: I would number the steps explicitly and make your comment about "most of these are plumbing" afterwards. Also the "decisions" should maybe be collapsible sections or even footnotes (especially "why LFortran?" this just seems too much of a diversion from me wanting to learn what Enzyme is and how it works).

Author: Disagree about "why LFortran" since it's really central to the stack (and has its own fanbase in the tech community).

(why-lfortran)=

> **Why LFortran.** The first step (`lfortran --show-llvm`) is also the reason the whole thing is tractable, since LFortran emits remarkably clean IR. Arrays come out as plain pointers with the usual GEP/load/store patterns, much like C, instead of the multi-field descriptor structs and runtime calls that Flang produces. That matters enormously, because Enzyme has to trace every memory access to figure out what's active, and it has a far easier time with pointer arithmetic than with opaque `fir.box` descriptors.

Reviewer: I would mention the obvious alternative (Flang) earlier so its clear what the choice is between.

; idx = (j - 1) * nx + i
%56 = sub i32 %55, 1                       ; j - 1
%58 = mul i32 %56, %57                      ; (j-1) * nx
%60 = add i32 %58, %59                      ; + i
store i32 %60, ptr %idx

; T_c = T_cur(idx)
%67 = getelementptr inbounds double, ptr %t_cur, i32 %65   ; &T_cur[idx]
%68 = load double, ptr %67                                 ; T_cur[idx]
store double %68, ptr %t_c

; kx_east = 2*(k0+k1*T_c)*(k0+k1*T_e) / ((k0+k1*T_c)+(k0+k1*T_e))
%110 = fmul double %108, %109              ; k1 * T_c
%111 = fadd double %107, %110              ; k0 + k1*T_c
%112 = fmul double 2.0, %111
; ... k0 + k1*T_e, the product, the sum ...
%130 = fdiv double %118, %129              ; the harmonic mean
store double %130, ptr %kx_east
```

That's it. Array access is a `getelementptr` + `load`, arithmetic is a flat sequence of `fmul`/`fadd`/`fdiv` that maps line-for-line onto the Fortran. Flang, by contrast, would wrap those same arrays in descriptor structs and reach for runtime helpers, and Enzyme would have to see through every one of them.

The catch is that LFortran is still maturing, so your code has to stay inside what it [supports](https://lfortran.org/progress/).

> **Why `-O1`, not `-O3`.** The `opt -O1` line above looks innocuous, and our first pipeline used `-O3` there. The forward pass worked perfectly, so we moved on. Later, the VJP returned NaN on certain inputs, and it took hours to find why. At `-O3`, LLVM's aggressive vectorization and code-motion produce IR patterns Enzyme apparently mishandles in reverse mode. In our case it bit when adjacent cell temperatures were equal, which caused intermediate terms cancel and a compiler rearrangement turned that into a division by zero. The fix is to keep optimization mild _before_ the AD pass and save `-O3` for _after_ it.

Reviewer: Do you need to tell a story here? This comes across as an LLM lab book. Answer the question directly rather than circle about

```suggestion
**Why `-O1`, not `-O3`.** At `-O3`, LLVM's aggressive vectorization and code-motion produce IR patterns Enzyme seems to mishandle in reverse mode leading to NaN's on certain inputs.For example, when adjacent cell temperatures are equal intermediate terms cancel and a compiler rearrangement turns that into division by zero. The fix is to keep optimization mild _before_ the AD pass and save `-O3` for _after_ it.
```

Author: We can massage the storytelling a bit but I do like the more personal tone.

> We certainly learned to test the gradients early and often.

Reviewer: Remove, you talk about this in the finite difference section.
```suggestion
```

  store double %916, ptr %"'de5"
  %917 = load double, ptr %"'de5"
  store double 0.0,  ptr %"'de5"
  %919 = load double, ptr %"'ipg6_unwrap"            ; read adjoint of T_init[idx] (slot of t_init')
  %920 = fadd fast double %919, %917                 ; propagate it upstream
  store double %920, ptr %"'ipg6_unwrap"
  ; ...
  %925 = add nsw i64 %924, -1                        ; loop counter runs DOWNWARD
  br label %invertloop.body
```

The loop induction variable counts _down_ (`add nsw i64 ..., -1`), and the body is the textbook reverse-mode dance: read an adjoint, zero it so it isn't double-counted, accumulate it into the upstream variable's shadow. Nobody wrote any of this — it's the chain rule, applied statement by statement to the solver's own IR. The full `@diffethermal_2d_solve` is about 6,900 lines. The 865-line input module comes out the other side at nearly 9,900.

(does-it-actually-work)=

## Does it actually work?

OK, it compiles. But after the `-O3` NaN incident, trusting it without checking is off the table. The safest test is to compare Enzyme's gradient against a ground-truth derivative we can compute _independently_, without Enzyme anywhere in the loop.

> For one regime we can do exactly that. If we set the conductivity's temperature coefficient $k_1 = 0$, the solver becomes linear: each explicit step is an affine map of the temperature field, $T^{s+1} = A(k_0)\,T^s + c(k_0)$, where the operators $A$ and $c$ encode the real stencil and the real mixed boundary conditions, and depend affinely on $k_0$. (Verified numerically: one step is affine in $k_0$ to a relative residual of ~1e-16.) That structure hands us the exact derivative of the _whole_ multi-step trajectory for free, via the tangent recurrence

Reviewer: ```suggestion
For one regime we can do exactly that. If we set the conductivity's temperature coefficient $k_1 = 0$, the solver becomes linear: each explicit step is an affine map of the temperature field, $T^{s+1} = A(k_0)\,T^s + c(k_0)$, where the operators $A$ and $c$ encode the real stencil and the real mixed boundary conditions, and are themselves affine in $k_0$. That structure hands us the exact derivative of the _whole_ multi-step trajectory for free, via the tangent recurrence
```

```bash
# Package the compiled solver as a served, differentiable JAX primitive
$ tesseract build demo/enzyme-lfortran/
$ tesseract serve enzyme-thermal-2d
```

The last piece is [Tesseract-JAX](https://github.com/pasteurlabs/tesseract-jax), which promotes a served Tesseract to a proper JAX primitive. Wrap a call in `apply_tesseract` and the Fortran solver becomes an ordinary differentiable JAX function: `jax.value_and_grad` routes its VJP calls straight through the HTTP boundary to the Enzyme-generated code, and the optimizer never has to know.

```python
# The Fortran solver, now callable as a differentiable JAX function
from tesseract_jax import apply_tesseract

def solve(T_init):
    inputs = {**base_inputs, "T_init": T_init}
    return apply_tesseract(enzyme_tess, inputs)["T_final"]  # differentiable Fortran!
```

With that in place, it's finally time to point a real optimization problem at it and find out whether all the effort actually paid off.

## Thermal forensics: recovering a 900-element initial temperature field

> Imagine a steel plate that went through some unmonitored heating event, maybe a laser pulse, maybe a localized defect, nobody knows. Five seconds later you get to measure temperatures at 100 sensor locations, and the question is whether you can reconstruct what the initial temperature distribution must have looked like.

Reviewer: nit: "must have looked like" feels a bit too strong in my opinion (you are after all making a regularization assumption)

```suggestion
Imagine a steel plate that went through some unmonitored heating event, maybe a laser pulse, maybe a localized defect, nobody knows. Five seconds later you get to measure temperatures at 100 sensor locations, and the question is whether you can reconstruct the initial temperature distribution.
```

> The true initial condition has two Gaussian hot spots sitting on a warm background. So the ask is 900 unknowns out of 100 noisy observations, run backwards through a nonlinear PDE. A small [Tikhonov term](https://en.wikipedia.org/wiki/Ridge_regression) penalizing departure from the ambient prior regularizes the otherwise ill-posed inversion, and since the loss is just a JAX function, that term is one extra line that `jax.value_and_grad` differentiates for free:

Reviewer: ```suggestion
The true initial condition has two Gaussian hot spots sitting on a warm background. So the ask is 900 unknowns out of 100 noisy observations, an underdetermined problem, run backwards through a nonlinear PDE. A small [Tikhonov term](https://en.wikipedia.org/wiki/Ridge_regression) penalizing departure from the ambient prior regularizes the otherwise ill-posed inversion, and since the loss is just a JAX function, that term is one extra line that `jax.value_and_grad` differentiates for free:
```

And because the solver is just another JAX function behind `apply_tesseract`, you could swap `enzyme_tess` out for a pure-JAX reimplementation tomorrow and nothing in the optimization loop would change, only the container sitting behind the HTTP call. Whether you find that abstraction beautiful or horrifying probably comes down to how you feel about your gradients quietly traversing six layers of indirection on the way through (Python → JAX/XLA → HTTP → ctypes → Enzyme → Fortran).

(whats-next-and-what-wed-do-differently)=

## What's next

A reality check before getting carried away: this is a 220-line solver written specifically to fit the pipeline. Enzyme itself handles loops, conditionals, and function calls just fine, and the Enzyme team has run it on much larger codes, including [BUDE molecular docking](https://enzyme.mit.edu/getting_started/UsingEnzyme/#bude) and [LULESH hydrodynamics](https://enzyme.mit.edu/getting_started/UsingEnzyme/#lulesh), with derivative overhead that usually lands somewhere between 1x and 4x. Still, "works on LULESH" and "works on your 30-year-old Fortran codebase" are two very different claims.

> We've flagged the sharp edges as we went, from the `_lfortran_malloc` crash to the `-O3` NaN incident to LFortran's still-incomplete coverage; the approach works, but it is nowhere near turnkey yet. Plan on adapting your Fortran, pinning your compiler versions, and dropping down to the IR level to debug at least once before you're done.

Reviewer: Strong AI smell here "to the `-O3` NaN incident."

```suggestion
We've flagged the sharp edges as we went, including array allocation not being traceable by Enzyme, the necessity of using `-O1` and LFortran's still-incomplete coverage; the approach works, but it is nowhere near turnkey yet. Plan on adapting your Fortran, pinning your compiler versions, and dropping down to the IR level to debug at least once before you're done.
```

> Two bigger questions stay open. One is **third-party code**: "but will it work on _my_ code?" is a perfectly fair thing to ask. The honest guess is that it probably will, after the kind of adaptation described earlier and likely a few surprises beyond it. The other is **scale**, where checkpointing schemes, MPI-parallel codes, GPU kernels, and multi-physics coupling all remain untested here. Enzyme has support for some of them, that's just territory this pipeline hasn't reached.

Reviewer: I think this has huge AI smell also "third-party code" and "_my_ code" are polar opposites as far as I interpret them. I'll let you address to ensure I don't change intended messaging.
